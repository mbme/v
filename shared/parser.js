/* eslint-disable no-labels, no-continue, no-extra-label, no-constant-condition, no-restricted-syntax */
// TODO preprocess: replace \r\n with \n
// TODO blockquote, code, list, secondary header

const isChar = char => (str, i) => str[i] === char
const isNewline = isChar('\n')

const declareType = type => ({ skip: [ 0, 0 ], children: [], isBreak: () => false, isValid: () => true, ...type })

export const Grammar = {
  Italic: declareType({
    skip: [ 1, 1 ],
    children: [ 'Bold' ],
    escapeChar: '_',
    isStart: (str, pos, context) => str[pos] === '_' && !context.includes('Italic'),
    isBreak: isNewline,
    isEnd: isChar('_'),
  }),

  Bold: declareType({
    skip: [ 1, 1 ],
    children: [ 'Italic' ],
    escapeChar: '*',
    isStart: (str, pos, context) => str[pos] === '*' && !context.includes('Bold'),
    isBreak: isNewline,
    isEnd: isChar('*'),
  }),

  Mono: declareType({
    skip: [ 1, 1 ],
    escapeChar: '`',
    isStart: isChar('`'),
    isBreak: isNewline,
    isEnd: isChar('`'),
  }),

  LinkName: declareType({
    skip: [ 1, 1 ],
    escapeChar: ']',
    isStart: isChar('['),
    isBreak: isNewline,
    isEnd: isChar(']'),
  }),

  LinkAddress: declareType({
    skip: [ 1, 0 ],
    escapeChar: ')',
    isStart: isChar('('),
    isBreak: isNewline,
    isEnd: isChar(')'),
  }),

  Link: declareType({
    skip: [ 0, 1 ],
    escapeChar: ')',
    children: [ 'LinkName', 'LinkAddress' ],
    isStart: isChar('['),
    isEnd: isChar(')'),
    isValid: ({ items }) => items.length === 2 && items[0].type === 'LinkName' && items[1].type === 'LinkAddress',
  }),

  Paragraph: declareType({
    children: [ 'Bold', 'Italic', 'Mono', 'Link' ],
    isStart: (str, pos) => pos === 0 || (str[pos] === '\n' && str[pos - 1] === '\n'),
    isEnd(str, pos) {
      if (pos === str.length) {
        return true
      }

      if (str[pos] !== '\n') {
        return false
      }


      if (pos + 1 === str.length || str[pos + 1] === '\n') {
        return true
      }

      return false
    },
  }),

  Header: declareType({
    skip: [ 1, 0 ],
    isStart: (str, pos) => {
      if (str[pos] !== '#' && str[pos + 1] !== ' ') {
        return false
      }

      if (pos > 0 && str[pos - 1] !== '\n') { // check newline before #
        return false
      }

      const newLinePos = str.indexOf('\n', pos)

      // check if there is an empty line after header
      if (newLinePos !== -1 && newLinePos + 1 < str.length && str[newLinePos + 1] !== '\n') {
        return false
      }

      return true
    },
    isEnd: (str, pos) => pos === str.length || str[pos] === '\n',
  }),

  Document: declareType({
    children: [ 'Header', 'Paragraph' ],
    isStart: (str, pos) => pos === 0,
    isEnd: (str, pos) => pos === str.length,
  }),
}

if (__DEVELOPMENT__) { // validate grammar
  Object.entries(Grammar).forEach(([ type, rule ]) => {
    rule.children.forEach(childType => !!Grammar[childType] || console.error(`WARN: ${type} has unknown child "${childType}"`))
  })
}

export function parseFrom(str, pos, type, context) {
  const rule = Grammar[type]
  const [ skipStart, skipEnd ] = rule.skip

  let i = pos
  if (!rule.isStart(str, i, context)) {
    return [ 0, null ]
  }

  i += skipStart

  const tree = {
    type,
    items: [],
  }
  let text = ''
  let ended = false

  outer:
  while (true) {
    // handle escapes
    if (str[i] === '\\' && rule.escapeChar === str[i + 1]) {
      text += str[i + 1]
      i += 2
      continue outer
    }

    if (rule.isEnd(str, i)) {
      ended = true
      i += skipEnd
      break outer
    }

    if (i === str.length) {
      break outer
    }

    inner:
    for (const childType of rule.children) {
      const [ length, leaf ] = parseFrom(str, i, childType, [ ...context, type ])
      if (!length) {
        continue inner
      }

      if (text) {
        tree.items.push(text)
        text = ''
      }

      i += length
      tree.items.push(leaf)

      continue outer
    }

    if (rule.isBreak(str, i)) {
      return [ 0, null ]
    }

    text += str[i]
    i += 1
  }

  if (text) {
    tree.items.push(text)
  }

  // validate result
  if (!ended || !rule.isValid(tree)) {
    return [ 0, null ]
  }

  const length = i - pos

  return [ length, tree ]
}

export default function parse(str, type) {
  const [ i, tree ] = parseFrom(str, 0, type, [])

  if (__DEVELOPMENT__ && i !== str.length) {
    console.error(`WARN: rule ${type} covers ${i} out of ${str.length} chars`)
  }

  return tree
}
