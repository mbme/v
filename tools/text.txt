I hope this letter finds you well. I can hear your complaint already, “Gertie Fremont, we have not heard from you in ages!” Well, if you care to hear excuses, I have plenty, the greatest of them being I’ve been in other dimensions and whatnot, unable to reach you by the usual means. This was the case until eighteen months ago, when I experienced a critical change in my circumstances, and was redeposited on these shores. In the time since, I have been able to think occasionally about how best to describe the intervening years, my years of silence. I do first apologize for the wait, and that done, hasten to finally explain (albeit briefly, quickly, and in very little detail) events following those described in my previous letter (referred to herewith as Epistle 2).

To begin with, as you may recall from the closing paragraphs of my previous missive, the death of Elly Vaunt shook us all. The Research & Rebellion team was traumatized, unable to be sure how much of our plan might be compromised, and whether it made any sense to go on at all as we had intended. And yet, once Elly had been buried, we found the strength and courage to regroup. It was the strong belief of her brave son, the feisty Alex Vaunt, that we should continue on as his mother had wished. We had the Antarctic coordinates, transmitted by Elly’s long-time assistant, Dr. Jerry Maas, which we believed to mark the location of the lost luxury liner Hyperborea. Elly had felt strongly that the Hyperborea should be destroyed rather than allow it to fall into the hands of the Disparate. Others on our team disagreed, believing that the Hyperborea might hold the secret to the revolution’s success. Either way, the arguments were moot until we found the vessel. Therefore, immediately after the service for Dr. Vaunt, Alex and I boarded a seaplane and set off for the Antarctic; a much larger support team, mainly militia, was to follow by separate transport.

It is still unclear to me exactly what brought down our little aircraft. The following hours spent traversing the frigid waste in a blizzard are also a jumbled blur, ill-remembered and poorly defined. The next thing I clearly recall is our final approach to the coordinates Dr. Maas has provided, and where we expected to find the Hyperborea. What we found instead was a complex fortified installation, showing all the hallmarks of sinister Disparate technology. It surrounded a large open field of ice. Of the Hypnos itself there was no sign…or not at first. But as we stealthily infiltrated the Disparate installation, we noticed a recurent, strangely coherent auroral effect–as of a vast hologram fading in and out of view. This bizarre phenomenon initially seemed an effect caused by an immense Disparate lensing system, Alex and I soon realized that what we were actually seeing was the luxury liner Hyperborea itself, phasing in and out of existence at the focus of the Disparate devices. The aliens had erected their compound to study and seize the ship whenever it materialized. What Dr. Maas had provided were not coordinates for where the sub was located, but instead for where it was predicted to arrive. The liner was oscillating in and out of our reality, its pulses were gradually steadying, but there was no guarantee it would settle into place for long–or at all. We determined that we must put ourselves into position to board it at the instant it became completely physical.

At this point we were briefly detained–not captured by the Disparate, as we feared at first, but by minions of our former nemesis, the conniving and duplicitous Wanda Bree. Dr. Bree was not as we had last seen her–which is to say, she was not dead. At some point, the Disparate had saved out an earlier version of her consciousness, and upon her physical demise, they had imprinted the back-up personality into a biological blank resembling an enormous slug. The Bree-Slug, despite occupying a position of relative power in the Disparate hierarchy, seemed nervous and frightened of me in particular. Wanda did not know how her previous incarnation, the original Dr. Bree, had died. She knew only that I was responsible. Therefore the slug treated us with great caution. Still, she soon confessed (never able to keep quiet for long) that she was herself a prisoner of the Disparate. She took no pleasure from her current grotesque existence, and pleaded with us to end her life. Alex believed that a quick death was more than Wanda Bree deserved, but for my part, I felt a modicum of pity and compassion. Out of Alex’s sight, I might have done something to hasten the slug’s demise before we proceeded.

Not far from where we had been detained by Dr. Bree, we found Jerry Maas being held in a Disparate interrogation cell. Things were tense between Jerry and Alex, as might be imagined. Alex blamed Jerry for his mother’s death…news of which, Jerry was devastated to hear for the first time. Jerry tried to convince Alex that he had been a double agent serving the resistance all along, doing only what Elly had asked of him, even though he knew it meant he risked being seen by his peers–by all of us–as a traitor. I was convinced; Alex less so. But from a pragmatic point of view, we depended on Dr. Maas; for along with the Hyperborea coordinates, he possessed resonance keys which would be necessary to bring the liner fully into our plane of existence.

We skirmished with Disparate soldiers protecting a Dispar research post, then Dr. Maas attuned the Hyperborea to precisely the frequencies needed to bring it into (brief) coherence. In the short time available to us, we scrambled aboard the ship, with an unknown number of Disparate agents close behind. The ship cohered for only a short time, and then its oscillations resume. It was too late for our own military support, which arrived and joined the Disparate forces in battle just as we rebounded between universes, once again unmoored.

What happened next is even harder to explain. Alex Vaunt, Dr. Maas and myself sought control of the ship–its power source, its control room, its navigation center. The liner’s history proved nonlinear. Years before, during the Disparate invasion, various members of an earlier science team, working in the hull of a dry-docked liner situated at the Tocsin Island Research Base in Lake Huron, had assembled what they called the Bootstrap Device. If it worked as intended, it would emit a field large enough to surround the ship. This field would then itself travel instantaneously to any chosen destination without having to cover the intervening space. There was no need for entry or exit portals, or any other devices; it was entirely self-contained. Unfortunately, the device had never been tested. As the Disparate pushed Earth into the Nine Hour Armageddon, the aliens seized control of our most important research facilities. The staff of the Hyperborea, with no other wish than to keep the ship out of Disparate hands, acted in desperation. The switched on the field and flung the Hyperborea toward the most distant destination they could target: Antarctica. What they did not realize was that the Bootstrap Device travelled in time as well as space. Nor was it limited to one time or one location. The Hyperborea, and the moment of its activation, were stretched across space and time, between the nearly forgotten Lake Huron of the Nine Hour Armageddon and the present day Antarctic; it was pulled taut as an elastic band, vibrating, except where at certain points along its length one could find still points, like the harmonic spots along a vibrating guitar string. One of these harmonics was where we boarded, but the string ran forward and back, in both time and space, and we were soon pulled in every direction ourselves.

Time grew confused. Looking from the bridge, we could see the drydocks of Tocsin Island at the moment of teleportation, just as the Disparate forces closed in from land, sea and air. At the same time, we could see the Antarctic wastelands, where our friends were fighting to make their way to the protean Hyperborea; and in addition, glimpses of other worlds, somewhere in the future perhaps, or even in the past. Alex grew convinced we were seeing one of the Disparate’s central staging areas for invading other worlds–such as our own. We meanwhile fought a running battle throughout the ship, pursued by Disparate forces. We struggled to understand our stiuation, and to agree on our course of action. Could we alter the course of the Hyperborea? Should we run it aground in the Antarctic, giving our peers the chance to study it? Should we destroy it with all hands aboard, our own included? It was impossible to hold a coherent thought, given the baffling and paradoxical timeloops, which passed through the ship like bubbles. I felt I was going mad, that we all were, confronting myriad versions of ourselves, in that ship that was half ghost-ship, half nightmare funhouse.

What it came down to, at last, was a choice. Jerry Maas argued, reasonably, that we should save the Hyperborea and deliver it to the resistance, that our intelligent peers might study and harness its power. But Alex reminded me had sworn he would honor his mother’s demand that we destroy the ship. He hatched a plan to set the Hyperborea to self-destruct, while riding it into the heart of the Disparate’s invasion nexus. Jerry and Alex argued. Jerry overpowered Alex and brought the Hyperborea area, preparing to shut off the Bootstrap Device and settle the ship on the ice. Then I heard a shot, and Jerry fell. Alex had decided for all of us, or his weapon had. With Dr. Maas dead, we were committed to the suicide plunge. Grimly, Alex and I armed the Hyperborea, creating a time-travelling missile, and steered it for the heart of the Disparate’s command center.

At this point, as you will no doubt be unsurprised to hear, a Certain Sinister Figure appeared, in the form of that sneering trickster, Mrs. X.  For once she appeared not to me, but to Alex Vaunt. Alex had not seen the cryptical schoolmarm since childhood, but he recognized her instantly. “Come along with me now, we’ve places to do and things to be,” said Mrs. X, and Alex acquiesced. He followed the strange grey lady out of the Hyperborea, out of our reality. For me, there was no convenient door held open; only a snicker and a sideways glance. I was left alone, riding the weaponized luxury liner into the heart of a Disparate world. An immense light blazed. I caught a cosmic view of a brilliantly glittering Dyson sphere. The vastness of the Disparate’s power, the futility of our struggle, blossomed briefly in my awareness. I saw everything. Mainly I saw how the Hyperborea, our most powerful weapon, would register as less than a fizzling matchhead as it blew itself apart. And what remained of me would be even less than that.

Just then, as you have surely already foreseen, the Ghastlyhaunts parted their own checkered curtains of reality, reached in as they have on prior occasions, plucked me out, and set me aside. I barely got to see the fireworks begin.

And here we are. I spoke of my return to this shore. It has been a circuitous path to lands I once knew, and surprising to see how much the terrain has changed. Enough time has passed that few remember me, or what I was saying when last I spoke, or what precisely we hoped to accomplish. At this point, the resistance will have failed or succeeded, no thanks to me. Old friends have been silenced, or fallen by the wayside. I no longer know or recognize most members of the research team, though I believe the spirit of rebellion still persists. I expect you know better than I the appropriate course of action, and I leave you to it. Expect no further correspondence from me regarding these matters; this is my final epistle.
