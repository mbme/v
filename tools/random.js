/* eslint-disable no-continue */
import { capitalize } from 'shared/utils'

// TODO handle few separators in a row

const SKIP_WORDS = [ 'dr.', 'mr.', 'mrs.' ]

function hasSkipWordAt(text, pos) {
  for (let i = 0; i < SKIP_WORDS.length; i += 1) {
    const word = SKIP_WORDS[i]
    if (pos + 1 < word.length) {
      continue
    }

    const skipWord = text.substring((pos + 1) - word.length, pos + 1).toLowerCase()
    if (skipWord === word) {
      return true
    }
  }

  return false
}

const isSeparator = char => char === '.' || char === '?' || char === '!'
const isPunctuation = char => char === ',' || char === ';' || char === ':'

export function getSentences(text) {
  const sentences = []

  let sentenceStart = 0
  for (let i = 0; i < text.length; i += 1) {
    if (isSeparator(text[i]) && !hasSkipWordAt(text, i)) {
      sentences.push(text.substring(sentenceStart, i + 1).trim())
      sentenceStart = i + 1
    }
  }

  if (!sentences.length) {
    sentences.push(text.trim())
  }

  return sentences
}

export function getWords(sentence) {
  const words = []

  let i = 0
  let word = ''
  while (i < sentence.length) {
    const char = sentence[i]

    if (char === ' ') {
      if (word.length) {
        words.push(word)
        word = ''
      }

      i += 1
      continue
    }

    if (isPunctuation(char) || (isSeparator(char) && !hasSkipWordAt(sentence, i))) {
      if (word.length) {
        words.push(word)
        word = ''
      }

      words.push(char)
      i += 1
      continue
    }

    word += char
    i += 1
  }

  if (word.length) {
    words.push(word)
  }

  return words.map(w => w.toLowerCase())
}

function calculateTextStats(text) {
  const dict = {}
  const separators = {}

  // 1. count words
  const sentences = getSentences(text)
  sentences.forEach((sentence) => {
    const words = getWords(sentence)

    const lastWord = words[words.length - 1]
    if (isSeparator(lastWord)) {
      separators[lastWord] = (separators[lastWord] || 0) + 1
      words.splice(words.length - 1, 1, lastWord.substring(0, lastWord.length - 1)) // remove separator from last word
    }

    words.forEach((word, index) => {
      const stats = dict[word] || { next: {}, nextWords: 0, start: 0, end: 0 }
      dict[word] = stats

      if (index === 0) {
        stats.start += 1
      }

      if (index === words.length - 1) {
        stats.end += 1
      }

      if (index < words.length - 1) {
        const nextWord = words[index + 1]
        stats.next[nextWord] = (stats.next[nextWord] || 0) + 1
        stats.nextWords += 1
      }
    })
  })

  // 2. calculate stats
  const stats = {
    starts: [],
    ends: [],
    separators: [],
    words: {},
  }

  Object.entries(dict).forEach(([ word, wordStats ]) => {
    if (wordStats.start > 0) {
      stats.starts.push([ word, wordStats.start / sentences.length ])
    }

    if (wordStats.end > 0) {
      stats.ends.push([ word, wordStats.end / sentences.length ])
    }


    stats.words[word] = Object.entries(wordStats.next).map(([ nextWord, usages ]) => [ nextWord, usages / wordStats.nextWords ])
  })
  Object.entries(separators).forEach(([ separator, usages ]) => {
    stats.separators.push([ separator, usages / sentences.length ])
  })

  return stats
}

/**
 * wordsDistribution: [[word, probability]]
 */
function pickWord(wordsDistribution) {
  if (!wordsDistribution.length) {
    return null
  }

  const wordProb = Math.random()
  let [ word, distribution ] = wordsDistribution[0]

  if (distribution > wordProb) {
    return word
  }

  for (let i = 1; i < wordsDistribution.length; i += 1) {
    word = wordsDistribution[i][0] // eslint-disable-line prefer-destructuring
    distribution += wordsDistribution[i][1]

    if (distribution > wordProb) {
      return word
    }
  }

  return word
}

function genSentence(stats) {
  let word = pickWord(stats.starts)
  let sentence = capitalize(word)

  let ended = pickWord(stats.ends) === word

  while (!ended) {
    word = pickWord(stats.words[word])

    if (!word) {
      ended = true
      continue
    }

    if (isPunctuation(word)) {
      sentence += word
    } else {
      sentence += ' ' + word
    }

    ended = pickWord(stats.ends) === word
  }

  return sentence + pickWord(stats.separators)
}

export function createTextGenerator(corpus) {
  const stats = calculateTextStats(corpus)

  return {
    sentence(minWords = 1, maxWords = 20, maxTries = 100) {
      let sentence = genSentence(stats)

      for (let tries = 0; tries < maxTries; tries += 1) {
        const words = sentence.split(' ')
        if (words.length >= minWords && words.length <= maxWords) {
          return sentence
        }

        sentence = genSentence(stats)
      }

      return sentence
    },
  }
}
