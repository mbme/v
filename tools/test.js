/* eslint-disable no-underscore-dangle */

import fs from 'fs';
import assert from 'assert';
import { readJSON, writeJSON } from 'core/utils';
import log from 'core/utils/log';
import { uniq } from 'shared/utils';

let _beforeCb;
let _tests = [];
let _afterCb;

export function collectTests(cb) {
  cb();

  const result = {
    tests: _tests,
    before: _beforeCb,
    after: _afterCb,
  };

  _beforeCb = null;
  _tests = [];
  _afterCb = null;

  return result;
}

export const test = (name, fn, only = false) => _tests.push({ name, fn, only });
export const before = (cb) => { _beforeCb = cb; };
export const after = (cb) => { _afterCb = cb; };

async function runTest({ name, fn }, oldSnapshots, updateSnapshots) {
  let okAsserts = 0;
  let snapshotPos = 0;
  const snapshots = [];

  try {
    await Promise.resolve(fn({
      equal(actual, expected) {
        if (actual === expected) {
          okAsserts += 1;
        } else {
          assert.fail(
            `not ok
            expected:
              ${expected}
            actual:
              ${actual}
          `
          );
        }
      },

      deepEqual(actual, expected) {
        assert.deepStrictEqual(actual, expected);
        okAsserts += 1;
      },

      matchSnapshot(actual) {
        if (snapshotPos < oldSnapshots.length) {
          try {
            assert.equal(
              JSON.stringify(actual, null, 2),
              JSON.stringify(oldSnapshots[snapshotPos], null, 2),
            );
          } catch (e) {
            if (!updateSnapshots) throw e;
            log.simple(`  ${name}: updating snapshot`);
          }
        }

        snapshots.push(actual);
        snapshotPos += 1;
        okAsserts += 1;
      },

      throws(block, error) {
        try {
          block();
          assert.fail('Expected to throw');
        } catch (e) {
          if (error) assert.equal(e, error);
        }
      },
    }));

    log.simple(`  ${name}: ${okAsserts} ok`, snapshotPos ? `/ ${snapshotPos} snapshots` : '');
    return [ snapshots, true ];
  } catch (e) {
    log.simple(`  ${name} failed\n`, e.message);
    log.simple(e.stack);
    return [ oldSnapshots, false ];
  }
}

export async function runTests(file, tests, updateSnapshots) {
  if (uniq(tests, ({ name }) => name).length !== tests.length) {
    throw new Error(`${file} contains tests with similar names`);
  }

  const snapshotsFile = file + '.snap.json';
  const snapshotsFileExists = fs.existsSync(snapshotsFile);
  const oldSnapshots = snapshotsFileExists ? await readJSON(snapshotsFile) : {};

  const newSnapshots = {};
  let failures = 0;

  for (const t of tests) {
    const [ snapshots, success ] = await runTest(t, oldSnapshots[t.name] || [], updateSnapshots);

    if (!success) failures += 1;

    if (snapshots.length) {
      newSnapshots[t.name] = snapshots;
    }
  }

  if (Object.values(newSnapshots).length) {
    await writeJSON(snapshotsFile, newSnapshots);
  } else if (snapshotsFileExists) {
    await fs.promises.unlink(snapshotsFile);
  }

  return failures;
}
