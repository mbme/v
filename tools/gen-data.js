import fs from 'fs';
import path from 'path';

import { createArray, randomInt, shuffle } from '../shared/utils';
import { createImageLink } from '../shared/parser';
import { readText, listFiles, sha256 } from '../core/utils';
import log from '../shared/log';
import { createTextGenerator } from './random';

async function listImages(basePath) {
  const files = await listFiles(basePath);
  const images = files.filter(name => name.match(/\.(jpg|jpeg)$/i));

  return Promise.all(images.map(async (name) => {
    const data = await fs.promises.readFile(path.join(basePath, name));
    const link = createImageLink(name, sha256(data));

    return { link, file: { name, data } };
  }));
}

async function genText(generator, images) {
  const name = generator.sentence(1, 8);

  const data = createArray(
    randomInt(1, 7), // paragraphs
    () => {
      const sentences = createArray(
        randomInt(1, 7), // sentences
        () => generator.sentence(),
      );

      if (Math.random() < 0.34) {
        const image = images[randomInt(0, images.length, false)];
        sentences.push(` ${image.link} `);
      }

      return shuffle(sentences).join(' ');
    }
  ).join('\n\n');

  return { name: name.substring(0, name.length - 1), data };
}

export default async function genData(api, notesCount) {
  const resourcesPath = path.join(__dirname, '../resources');
  const images = await listImages(resourcesPath);
  const text = await readText(path.join(resourcesPath, 'text.txt'));
  const generator = createTextGenerator(text);

  const notesPromises = createArray(notesCount, async () => {
    const { name, data } = await genText(generator, images);
    return api.CREATE_NOTE({ name, data }, images.map(image => image.file.data));
  });

  await Promise.all(notesPromises);

  log.info('Generated %s fake notes', notesCount);
}
