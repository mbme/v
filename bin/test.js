/* eslint-disable import/no-dynamic-require, global-require, no-await-in-loop */
import path from 'path'
import { walkSync } from 'server/utils'
import { collectTests, runTests } from 'tools/test'

const filter = process.argv.length > 3 ? process.argv[3] : ''
const updateSnapshots = process.env.UPDATE_SNAPSHOTS === 'true'

const basePath = path.join(__dirname, '..')
const testFiles = walkSync(basePath)
  .map(filePath => path.relative(basePath, filePath))
  .filter(relPath => relPath.endsWith('.test.js') && relPath.includes(filter))

const testPlans = []
for (const testFile of testFiles) {
  const testPlan = collectTests(() => require(testFile))
  const only = testPlan.tests.find(test => test.only)
  if (only) {
    if (updateSnapshots) throw new Error("Can't update the 'only' snapshot")

    testPlans.length = 0
    testPlans.push({ file: testFile, ...testPlan, tests: [ only ] })
    break
  } else {
    testPlans.push({ file: testFile, ...testPlan })
  }
}

async function executeTestPlans() {
  for (const testPlan of testPlans) {
    console.log(testPlan.file)

    testPlan.before && await Promise.resolve(testPlan.before())
    await runTests(path.join(basePath, testPlan.file), testPlan.tests, updateSnapshots)
    testPlan.after && await Promise.resolve(testPlan.after())

    console.log('')
  }
}

executeTestPlans()
