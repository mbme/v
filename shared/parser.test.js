import { test, before, after } from 'tools/test'
import { parse, select, selectLinks } from './parser'

const text = `
# Header

block
# so not a header

* test
* test

Paragraph and something else. sentence
test *bold* \`code\`

\`\`\`js
 code block
\`\`\`

> some blockquote
> quote

One more paragraph. [link](http://link.to/123?321)

`

before(() => { global.__DEVELOPMENT__ = false }) // suppress warnings
after(() => { global.__DEVELOPMENT__ = true })

test('Bold', (assert) => {
  assert.matchSnapshot(parse('*test*', 'Bold'))
  assert.matchSnapshot(parse('* test \\* and* so on', 'Bold'))
  assert.equal(parse(
    `*te
    st*`,
    'Bold',
  ), null)
})

test('Mono', (assert) => {
  assert.matchSnapshot(parse('`test`', 'Mono'))
  assert.matchSnapshot(parse('` test \\` and` so on', 'Mono'))
  assert.matchSnapshot(parse(
    `\`te
    st\``,
    'Mono',
  ), null)
})

test('Header', (assert) => {
  assert.matchSnapshot(parse('# AHAHAH *test oh no* !', 'Header'))
  assert.equal(parse(
    `# AHAHAH *test oh no* !
    there is no empty line
    `,
    'Header',
  ), null)
})

test('Link', (assert) => {
  assert.matchSnapshot(parse('[awesome link](http://amazing.com)', 'Link'))
})

test('Paragraph', (assert) => {
  assert.matchSnapshot(parse(
    `
    AHAHAH *test oh no!
    go go _power cows_ \`code\`
    `,
    'Paragraph',
  ))
})

test('Document', (assert) => {
  assert.matchSnapshot(parse(text, 'Document'))
})

test('select', (assert) => {
  const links = select(parse(text), 'Link')
  assert.matchSnapshot(links)
})

test('selectLinks', (assert) => {
  const links = selectLinks(parse(text))
  assert.matchSnapshot(links)
})
