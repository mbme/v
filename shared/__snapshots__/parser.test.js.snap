// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser Bold 1`] = `
Object {
  "items": Array [
    "test",
  ],
  "type": "Bold",
}
`;

exports[`Parser Bold 2`] = `
Object {
  "items": Array [
    " test * \\\\ ",
    Object {
      "items": Array [
        "and*",
      ],
      "type": "Italic",
    },
    " so on",
  ],
  "type": "Bold",
}
`;

exports[`Parser Document 1`] = `
Object {
  "items": Array [
    Object {
      "items": Array [
        "
# Header",
      ],
      "type": "Paragraph",
    },
    "
",
    Object {
      "items": Array [
        "
block
# so not a header",
      ],
      "type": "Paragraph",
    },
    "
",
    Object {
      "items": Array [
        "
* test
* test",
      ],
      "type": "Paragraph",
    },
    "
",
    Object {
      "items": Array [
        "
Paragraph and something else. sentence
test ",
        Object {
          "items": Array [
            "bold",
          ],
          "type": "Bold",
        },
        " ",
        Object {
          "items": Array [
            "italic",
          ],
          "type": "Italic",
        },
        " ",
        Object {
          "items": Array [
            Object {
              "items": Array [
                "bold italic",
              ],
              "type": "Bold",
            },
          ],
          "type": "Italic",
        },
        " ",
        Object {
          "items": Array [
            "code",
          ],
          "type": "Mono",
        },
      ],
      "type": "Paragraph",
    },
    "
",
    Object {
      "items": Array [
        "
",
        Object {
          "items": Array [],
          "type": "Mono",
        },
        "\`js
 code block
",
        Object {
          "items": Array [],
          "type": "Mono",
        },
        "\`",
      ],
      "type": "Paragraph",
    },
    "
",
    Object {
      "items": Array [
        "
> some blockquote
> quote",
      ],
      "type": "Paragraph",
    },
    "
",
    Object {
      "items": Array [
        "
One more paragraph. ",
        Object {
          "items": Array [
            Object {
              "items": Array [
                "link",
              ],
              "type": "LinkName",
            },
            Object {
              "items": Array [
                "http://link.to/123?321",
              ],
              "type": "LinkAddress",
            },
          ],
          "type": "Link",
        },
      ],
      "type": "Paragraph",
    },
    "

",
  ],
  "type": "Document",
}
`;

exports[`Parser Header 1`] = `
Object {
  "items": Array [
    " AHAHAH *test oh no* !",
  ],
  "type": "Header",
}
`;

exports[`Parser Italic 1`] = `
Object {
  "items": Array [
    "test",
  ],
  "type": "Italic",
}
`;

exports[`Parser Italic 2`] = `
Object {
  "items": Array [
    " test _ and so on ",
    Object {
      "items": Array [
        "_test",
      ],
      "type": "Bold",
    },
  ],
  "type": "Italic",
}
`;

exports[`Parser Link 1`] = `
Object {
  "items": Array [
    Object {
      "items": Array [
        "awesome link",
      ],
      "type": "LinkName",
    },
    Object {
      "items": Array [
        "http://amazing.com",
      ],
      "type": "LinkAddress",
    },
  ],
  "type": "Link",
}
`;

exports[`Parser Mono 1`] = `
Object {
  "items": Array [
    "test",
  ],
  "type": "Mono",
}
`;

exports[`Parser Mono 2`] = `
Object {
  "items": Array [
    " test \` \\\\ _and*_ so on*",
  ],
  "type": "Mono",
}
`;

exports[`Parser Paragraph 1`] = `
Object {
  "items": Array [
    "
      AHAHAH ",
    Object {
      "items": Array [
        "test oh no",
      ],
      "type": "Bold",
    },
    " !
      go go ",
    Object {
      "items": Array [
        "power cows",
      ],
      "type": "Italic",
    },
    " ",
    Object {
      "items": Array [
        "code",
      ],
      "type": "Mono",
    },
    "
      ",
  ],
  "type": "Paragraph",
}
`;
